### DISCUSSION TITLE
E-commerce datamodel question

### DISCUSSION THREAD

**Discussion by kees-tce** [directus/directus#4894] [Archive]
Hi all!

Really interested in a few opinions on the following.

I've added 3 collections:
1. Suppliers
2. Products
3. ProductSuppliers (which is a many to many relation table)

A product has a price and stock per supplier.

Two questions:
1. Any thoughts/experience on similar data models?
2. Let's say I want to update the stock and price regularly based on endpoints/csv's from the suppliers. What's the best extension in Directus? And how do I update objects in the ProductSuppliers collection best? Through Api, graph or database sql? Keep in mind that with the api and graph api I need to pass the PK, which I don't know from the supplier input.

Thanks for any input and ideas!

Kees.

**Comment by rijkvanzanten** [comment]
> Any thoughts/experience on similar data models?

Question: In your project, can a single product be supplied by multiple suppliers? Other than that, I see no problems with it

> Let's say I want to update the stock and price regularly based on endpoints/csv's from the suppliers. What's the best extension in Directus?

That sounds like a custom endpoint to me. You'd be able to process the data in the endpoint and inject it where it needs to go directly.

> And how do I update objects in the ProductSuppliers collection best? Through Api, graph or database sql? Keep in mind that with the api and graph api I need to pass the PK, which I don't know from the supplier input.

That's a little hard to tell without knowing exactly what sort of updates you're looking to do. 

You'd want to have some sort of way to identify the suppliers input data when connecting it to Directus, regardless of how you end up updating the data in the DB. If this external CSV of data has it's own PK for the supplier, you could use that as the primary key for `Suppliers` (using the manual string PK) and update it based on that

**Reply by keesvanbemmel** [reply]
Interestingly enough, I'm on a similar path. A little different though.

In the ProductSuppliers table, I have an ID, a Supplier ID and a supplier_product_id. The latter is unique for a supplier, but chances are supplier "a" has a different product with id "supplier-id-1" than supplier "b".

So what I thought of doing, is just update all rows using the knex instance while iterating through the CSV. Set stock/price where supplier id = 1 and supplier_product_id = product_id_from_csv[i].

I'm now at the point where I have registered a custom API hook and listening to a file creation (which will be the CSV).

Next step, and I think I need some pointers, is using the FilesService in that hook to get the contents from that CSV. Are there any examples I can borrow some knowledge of? Same question for the database knex instance in the hook.

**Reply by fitu86** [reply]
https://discord.com/channels/725371605378924594/726095500964528199/902970193372938310
